aten/src/ATen/native/xnnpack/Convolution.cpp

```diff
@@ -0,0 +1,302 @@
+      dilation[Layout::Parameter::Height],                            // dilation_height
+      dilation[Layout::Parameter::Width],                             // dilation_width
+      groups,                                                         // groups
+      weight_nhwc.size(Layout::Filter::Input),                        // group_input_channels
+      weight_nhwc.size(Layout::Filter::Output) / groups,              // group_output_channels
+      weight_nhwc.size(Layout::Filter::Input) * groups,               // input_pixel_stride
+      weight_nhwc.size(Layout::Filter::Output),                       // output_pixel_stride
+      weight_nhwc.data_ptr<float>(),                                  // kernel
+      (bias && bias->defined()) ? bias->data_ptr<float>() : nullptr,  // bias
+      output_min.to<float>(),                                         // output_min
+      output_max.to<float>(),                                         // output_max
+      0u,                                                             // flags
+      &convolution_op);                                               // operator
+
+  TORCH_INTERNAL_ASSERT(
+      xnn_status_success == create_status,
+      "xnn_create_convolution2d_nhwc_f32 failed!");
+
+  const auto Spatial = [&](const size_t filter, const size_t parameter) {
+    const int64_t kernel = 1 + dilation[parameter] * (weight_nhwc.size(filter) - 1);
+
+    return Context::Output::Spatial{
+      stride[parameter] - kernel + 2 * padding[parameter],
+      stride[parameter],
+    };
+  };
+
+  return Context{
+    Operator(convolution_op),
+    {
+      weight_nhwc.size(Layout::Filter::Output),
+      Spatial(Layout::Filter::Height, Layout::Parameter::Height),
+      Spatial(Layout::Filter::Width, Layout::Parameter::Width),
+    },
+  };
+}
+
+bool usable(const Tensor& input) {
+         // Input
+  return (4 == input.ndimension()) &&
+         (c10::DeviceType::CPU == input.device().type()) &&
+         (kFloat == input.scalar_type()) &&
+         (input.size(Layout::Activation4D::Batch) > 0) &&
+         (input.size(Layout::Activation4D::Channels) > 0) &&
+         (input.size(Layout::Activation4D::Height) > 0) &&
+         (input.size(Layout::Activation4D::Width) > 0);
+}
+
+Tensor run(
+    const Context& context,
+    const Tensor& input) {
+  using namespace internal;
+
+  const Tensor input_nhwc = input.contiguous(MemoryFormat::ChannelsLast);
+
+  TORCH_CHECK(
+      usable(input_nhwc),
+      "xnnpack::convolution not usable!");
+
+  const auto Spatial = [](const int64_t input, const Context::Output::Spatial spatial) {
+    return (input + spatial.alpha) / spatial.beta;
+  };
+
+  Tensor output = empty(
+      std::array<int64_t, 4>{
+        input_nhwc.size(Layout::Activation4D::Batch),
+        context.output.channels,
+        Spatial(input_nhwc.size(Layout::Activation4D::Height), context.output.height),
```

Comment:
Recently we unified computing output shape, so that now all backends call the same function https://github.com/pytorch/pytorch/blob/5380e16db93358e169dcce90c49e589c1ce67198/aten/src/ATen/native/ConvUtils.h#L25-L42. It would be awesome if xnnpack could do the same, having shape propagation in different places makes the codebase harder to understand. Formulas are actually the same, I checked :-)

Reply:
Will do.